Don't show App Store link previews when network connection integrity is enabled

rr-codes opened the pull request:
#### 12df7c45b790c57f59631375a2d59ecbc6f1d4ec
<pre>
Don&apos;t show App Store link previews when network connection integrity is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=256723">https://bugs.webkit.org/show_bug.cgi?id=256723</a>
rdar://103849151

Reviewed by Wenson Hsieh.

If network connection integrity is enabled, and if the link is an App Store or iTunes link,
do not show the link preview.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView continueContextMenuInteraction:]):

Canonical link: <a href="https://commits.webkit.org/264553@main">https://commits.webkit.org/264553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bc11f7cdcdc85276000b54f0b67274e3f4c9ded

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7924 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9298 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7834 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7849 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10700 "1 flakes 98 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7047 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9408 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6233 "Passed tests") | | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14659 "6 flakes 110 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10512 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7591 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6192 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6924 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1900 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11135 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12317,7 +12317,11 @@ - (void)continueContextMenuInteraction:(void(^)(UIContextMenuConfiguration *))co
 
 #if ENABLE(DATA_DETECTION)
     if ([(NSURL *)linkURL iTunesStoreURL]) {
-        [self continueContextMenuInteractionWithDataDetectors:continueWithContextMenuConfiguration];
```
Can we still show a standard link preview in this case?

rr-codes replied:
According to https://bugs.webkit.org/show_bug.cgi?id=200485 , we actually can't

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5703,6 +5703,8 @@ void WebPageProxy::didCommitLoadForFrame(FrameIdentifier frameID, FrameInfoData&
 #if ENABLE(NETWORK_CONNECTION_INTEGRITY)
         updateAllowedLookalikeCharacterStringsIfNeeded();
+        if (navigation && navigation->websitePolicies())
+            m_networkConnectionIntegrityPolicies = navigation->websitePolicies()->networkConnectionIntegrityPolicy();
```
Nit - can we also reset `m_networkConnectionIntegrityPolicies` in `WebPageProxy::resetState`?

charliewolfe commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12317,7 +12317,11 @@ - (void)continueContextMenuInteraction:(void(^)(UIContextMenuConfiguration *))co
     if ([(NSURL *)linkURL iTunesStoreURL]) {
-        [self continueContextMenuInteractionWithDataDetectors:continueWithContextMenuConfiguration];
+#if ENABLE(NETWORK_CONNECTION_INTEGRITY)
+        if (!_page->networkConnectionIntegrityPolicies().contains(WebCore::NetworkConnectionIntegrity::Enabled))
```
Are we sure this is what we want to check? Instead of checking for an ephemeral session?

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12317,7 +12317,21 @@ - (void)continueContextMenuInteraction:(void(^)(UIContextMenuConfiguration *))co
+#endif
+
+            return true;
+        }();
```
(optional) Nit: since we're only using `_page` maybe just capture that?

```suggestion
        bool shouldContinueContextMenuInteraction = [&page = _page] {
            if (!page->websiteDataStore().isPersistent())
                return false;

#if ENABLE(NETWORK_CONNECTION_INTEGRITY)
            if (page->networkConnectionIntegrityPolicies().contains(WebCore::NetworkConnectionIntegrity::Enabled))
                return false;
#endif

            return true;
        }();
```

whsieh replied:
I think either approach seems reasonable. This check also seems simple enough, that you could consider removing the temporary variable altogether and just do:

```
if (_page->websiteDataStore().isPersistent() && !_page->networkConnectionIntegrityPolicies().contains(WebCore::NetworkConnectionIntegrity::Enabled))
    [self continueContextMenuInteractionWithDataDetectors:continueWithContextMenuConfiguration];
else
    continueWithContextMenuConfiguration(nil);
return;
```

rr-codes replied:
@whsieh what about the compile time guards though?

whsieh replied:
Oh, right. Something like this?

```
    bool shouldUseDataDetectors = _page->websiteDataStore().isPersistent();
#if ENABLE(NETWORK_CONNECTION_INTEGRITY)
    if (_page->networkConnectionIntegrityPolicies().contains(WebCore::NetworkConnectionIntegrity::Enabled))
        shouldUseDataDetectors = false;
#endif
    …
```

(Of course, lambda is fine too. Your call!)

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12317,7 +12317,23 @@ - (void)continueContextMenuInteraction:(void(^)(UIContextMenuConfiguration *))co
+            [self continueContextMenuInteractionWithDataDetectors:continueWithContextMenuConfiguration];
+        else
+            continueWithContextMenuConfiguration(nil);
+
```
Nit - extra newline.

rr-codes replied:
is this really an extra newline? I did it intentionally since otherwise its immediately after the `else` branch which makes it harder to read

rr-codes committed:
Don't show App Store link previews when network connection integrity is enabled
https://bugs.webkit.org/show_bug.cgi?id=256723
rdar://103849151

Reviewed by Wenson Hsieh.

If network connection integrity is enabled, and if the link is an App Store or iTunes link,
do not show the link preview.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView continueContextMenuInteraction:]):

Canonical link: https://commits.webkit.org/264553@main
